src/main/java/edu/usf/cutr/tba/manager/TravelBehaviorDataAnalysisManager.java

```diff
@@ -239,10 +260,99 @@ private void completeTravelBehaviorRecord(TravelBehaviorInfo tbi, QueryDocumentS
             mLastTravelBehaviorRecord.setOriginDestinationDistance(distance);
         }
 
-        String regionId = findRegionIdFromDeviceInfoList(doc, userDeviceInfoList);
-        mLastTravelBehaviorRecord.setRegionId(regionId);
+        //String regionId = findRegionIdFromDeviceInfoList(doc, userDeviceInfoList);
+        //mLastTravelBehaviorRecord.setRegionId(regionId);
+
+        //While completing TBR update isIgnoringBatteryOptimization, isTalkBackEnabled
+        //and isPowerSaveModeOn
+        if(mLastTravelBehaviorRecord.getActivityEndTimeMillis() != null){ /*if valid activity end time*/
+            //Get the previous DeviceInformation but closest in time to the activity end time
+            DeviceInformation dvInfo = getClosestDeviceInfo(userDeviceInfoList);
+            if(dvInfo != null) { /*If there is DeviceInformation available*/
+                //Get the corresponding properties when available
+                if(dvInfo.getIgnoringBatteryOptimizations() != null){
+                    mLastTravelBehaviorRecord.setIsIgnoringBatteryOptimization(dvInfo.getIgnoringBatteryOptimizations());
+                }
+                if(dvInfo.getTalkBackEnabled() != null){
+                    mLastTravelBehaviorRecord.setIsTalkBackEnabled(dvInfo.getTalkBackEnabled());
+                }
+
+                if(dvInfo.getPowerSaveModeEnabled() != null){
+                    mLastTravelBehaviorRecord.setIsPowerSaveModeEnabled(dvInfo.getPowerSaveModeEnabled());
+                }
+                //setRegionID
+                mLastTravelBehaviorRecord.setRegionId(String.valueOf(dvInfo.regionId));
+            }
+        }
+    }
+
+    /**
+     * Performs a binary search to return the DeviceInformation object which timestamp is closest to the
+     * activityEndTime.
+     * @param userDeviceInfoList Sorted list of QueryDocumentSnapshot including information over time of user device
+     * @return DeviceInformation object with the timestamp closest to the activityEndTime. If the
```

Comment:
"closest" here means object with the nearest timestamp that occurs prior to (and not after) the activityEndTime, right? If so please update the Javadocs to clarify this.